File: src/MagnusAddon/language_services/jamdict_ex/jamdict_threading_wrapper.py
from __future__ import annotations

import queue
import threading
from concurrent.futures import Future
from typing import TYPE_CHECKING, Any

from ankiutils import app
from autoslot import Slots
from jamdict import Jamdict  # pyright: ignore [reportMissingTypeStubs]
from sysutils.lazy import Lazy
from sysutils.typed import non_optional, str_
from typed_linq_collections.collections.q_list import QList

if TYPE_CHECKING:
    from collections.abc import Callable

    from jamdict.util import LookupResult  # pyright: ignore [reportMissingTypeStubs]

class Request[T](Slots):
    def __init__(self, func: Callable[[Jamdict], T], future: Future[T]) -> None:
        self.func: Callable[[Jamdict], T] = func
        self.future: Future[T] = future

class JamdictThreadingWrapper(Slots):
    def __init__(self) -> None:
        self._queue: queue.Queue[Request[Any]] = queue.Queue()  # pyright: ignore[reportExplicitAny]
        self._thread: threading.Thread = threading.Thread(target=self._worker, daemon=True)
        self._running: bool = True
        self._thread.start()
        self.jamdict: Lazy[Jamdict] = Lazy(self.create_jamdict)

    @staticmethod
    def create_jamdict() -> Jamdict:
        return Jamdict(memory_mode=True) \
            if (app.config().load_jamdict_db_into_memory.get_value()
                and not app.is_testing) \
            else Jamdict(reuse_ctx=True)

    def _worker(self) -> None:
        while self._running:
            request = self._queue.get()
            try:
                result = request.func(self.jamdict())  # pyright: ignore[reportAny]
                request.future.set_result(result)
            except Exception as e:
                request.future.set_exception(e)

    def lookup(self, word: str, include_names: bool) -> LookupResult:
        future: Future[LookupResult] = Future()

        def do_actual_lookup(jamdict: Jamdict) -> LookupResult:
            return jamdict.lookup(word, lookup_chars=False, lookup_ne=include_names)  # pyright: ignore[reportUnknownMemberType]

        self._queue.put(Request(do_actual_lookup, future))
        return future.result()

    def run_string_query(self, sql_query: str) -> QList[str]:
        def perform_query(jamdict: Jamdict) -> QList[str]:
            result: QList[str] = QList()
            for batch in non_optional(non_optional(jamdict.jmdict).ctx().conn).execute(sql_query):  # pyright: ignore[reportAny]
                for row in batch:  # pyright: ignore[reportAny]
                    result.append(str_(row))  # noqa: PERF401  # pyright: ignore[reportAny]

            return result

        future: Future[QList[str]] = Future()

        self._queue.put(Request(perform_query, future))
        return future.result()

File: src/src_dotnet/JAStudio.Core/LanguageServices/JamdictEx/JamdictThreadingWrapper.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using JAStudio.Core.Configuration;
using JAStudio.PythonInterop.Utilities;
using Python.Runtime;

namespace JAStudio.Core.LanguageServices.JamdictEx;

class Request<T>
{
   public Func<dynamic, T> Func { get; }
   public TaskCompletionSource<T> TaskCompletionSource { get; }

   public Request(Func<dynamic, T> func, TaskCompletionSource<T> taskCompletionSource)
   {
      Func = func;
      TaskCompletionSource = taskCompletionSource;
   }
}

class JamdictLookupResult
{
   public List<DictEntry> Entries { get; }
   public List<DictEntry> Names { get; }

   public JamdictLookupResult(List<DictEntry> entries, List<DictEntry> names)
   {
      Entries = entries;
      Names = names;
   }
}

class JamdictThreadingWrapper : IDisposable
{
   static readonly object Lock = new();
   static JamdictThreadingWrapper? _instance;

   public static JamdictThreadingWrapper GetInstance(JapaneseConfig config)
   {
      if(_instance != null) return _instance;
      lock(Lock)
      {
         return _instance ??= new JamdictThreadingWrapper(config);
      }
   }

   public static void ShutDown()
   {
      lock(Lock)
      {
         _instance?.Dispose();
         _instance = null;
      }
   }

   readonly BlockingCollection<object> _queue;

   // ReSharper disable once PrivateFieldCanBeConvertedToLocalVariable
   readonly Thread _thread;
   dynamic? _jamdict;
   volatile bool _running;
   readonly JapaneseConfig _config;

   JamdictThreadingWrapper(JapaneseConfig config)
   {
      _config = config;
      _running = true;
      _queue = new BlockingCollection<object>();
      _thread = new Thread(Worker) { IsBackground = true };
      _thread.Start();
   }

   public void Dispose()
   {
      _running = false;
      _queue.CompleteAdding();
      _thread.Join(TimeSpan.FromSeconds(5));
      _queue.Dispose();
   }

   dynamic CreateJamdict()
   {
      // NOTE: Caller must already hold the GIL
      dynamic jamdictModule = Py.Import("jamdict");

      var memoryMode = _config.LoadJamdictDbIntoMemory.Value && !CoreApp.IsTesting;

      if(memoryMode)
      {
         return jamdictModule.Jamdict(memory_mode: true);
      } else
      {
         return jamdictModule.Jamdict(reuse_ctx: true);
      }
   }

   dynamic GetJamdict()
   {
      // NOTE: Caller must already hold the GIL
      if(_jamdict == null)
      {
         _jamdict = CreateJamdict();
      }

      return _jamdict;
   }

   void Worker()
   {
      while(_running)
      {
         try
         {
            // Wait for request WITHOUT holding GIL
            var request = _queue.Take();

            // Acquire GIL only for processing this request
            using(PythonEnvironment.LockGil())
            {
               try
               {
                  if(request is Request<JamdictLookupResult> lookupRequest)
                  {
                     var result = lookupRequest.Func(GetJamdict());
                     lookupRequest.TaskCompletionSource.SetResult(result);
                  } else if(request is Request<List<string>> stringListRequest)
                  {
                     var result = stringListRequest.Func(GetJamdict());
                     stringListRequest.TaskCompletionSource.SetResult(result);
                  }
               }
               catch(Exception e)
               {
                  if(request is Request<JamdictLookupResult> lookupRequest)
                  {
                     lookupRequest.TaskCompletionSource.SetException(e);
                  } else if(request is Request<List<string>> stringListRequest)
                  {
                     stringListRequest.TaskCompletionSource.SetException(e);
                  }
               }
            }
         }
         catch(Exception)
         {
            // Queue closed or other error
            if(!_running) break;
         }
      }
   }

   public JamdictLookupResult Lookup(string word, bool includeNames)
   {
      var tcs = new TaskCompletionSource<JamdictLookupResult>();

      JamdictLookupResult DoActualLookup(dynamic jamdict)
      {
         var lookupResult = jamdict.lookup(word, lookup_chars: false, lookup_ne: includeNames);

         // Convert to .NET types while holding GIL
         var entries = DictEntry.CreateFromPythonEntries(lookupResult.entries);
         var names = DictEntry.CreateFromPythonEntries(lookupResult.names);

         return new JamdictLookupResult(entries, names);
      }

      var request = new Request<JamdictLookupResult>(DoActualLookup, tcs);
      _queue.Add(request);

      return tcs.Task.Result;
   }

   public List<string> RunStringQuery(string sqlQuery)
   {
      List<string> PerformQuery(dynamic jamdict)
      {
         var result = new List<string>();

         var conn = jamdict.jmdict.ctx().conn;
         foreach(var batch in Dyn.Enumerate(conn.execute(sqlQuery)))
         {
            foreach(var row in Dyn.Enumerate(batch))
            {
               result.Add((string)row);
            }
         }

         return result;
      }

      var tcs = new TaskCompletionSource<List<string>>();
      var request = new Request<List<string>>(PerformQuery, tcs);
      _queue.Add(request);

      return tcs.Task.Result;
   }
}
